plat/kvm/efi.c

```diff
@@ -0,0 +1,342 @@
+ * the MAT also differentiates between permissions of the Runtime Services'
+ * PE sections (Runtime Services can basically be thought of as loaded Portable
+ * Executable format drivers).
+ */
+static void uk_efi_rt_md_to_bi_mrds(struct ukplat_memregion_desc **rt_mrds,
+				    __u32 *const rt_mrds_count)
+{
+	struct ukplat_memregion_desc *rt_mrd;
+	uk_efi_mem_attr_tab_t *mat = NULL;
+	uk_efi_mem_desc_t *mat_md;
+	uk_efi_status_t status;
+	uk_efi_cfg_tab_t *ct;
+	uk_efi_uintn_t i;
+	__sz desc_sz;
+
+	/* Search for the MAT in UEFI System Table's Configuration Tables */
+	for (i = 0; i < uk_efi_st->number_of_table_entries; i++) {
+		ct = &uk_efi_st->configuration_table[i];
+
+		if (!memcmp(&ct->vendor_guid,
+			    UK_EFI_MEMORY_ATTRIBUTES_TABLE_GUID,
+			    sizeof(ct->vendor_guid))) {
+			mat = ct->vendor_table;
+			break;
+		}
+	}
+	if (unlikely(!mat))
+		uk_efi_crash("Could not find Memory Attribute Table\n");
+
+	desc_sz = mat->descriptor_size;
+	*rt_mrds_count = mat->number_of_entries;
+	status = uk_efi_bs->allocate_pool(UK_EFI_LOADER_DATA,
+					  *rt_mrds_count * sizeof(**rt_mrds),
+					  (void **)rt_mrds);
+	if (unlikely(status != UK_EFI_SUCCESS))
+		uk_efi_crash("Failed to allocate memory for Memory Sub-region Descriptors\n");
+
+	/* Convert the EFI Runtime Services Memory descriptors to
+	 * ukplat_memregion_desc's
+	 */
+	mat_md = (uk_efi_mem_desc_t *)mat->entry;
+	for (i = 0; i < *rt_mrds_count; i++) {
+		if (!(mat_md->attribute & UK_EFI_MEMORY_RUNTIME))
+			continue;
+
+		rt_mrd = *rt_mrds + i;
+		rt_mrd->pbase = mat_md->physical_start;
+		rt_mrd->len = mat_md->number_of_pages * UK_EFI_PAGE_SIZE;
+		rt_mrd->vbase = rt_mrd->pbase;
+		rt_mrd->type = UKPLAT_MEMRT_RESERVED;
+		rt_mrd->flags = UKPLAT_MEMRF_MAP;
+		if (mat_md->attribute & UK_EFI_MEMORY_XP)
+			if (mat_md->attribute & UK_EFI_MEMORY_RO)
+				rt_mrd->flags |= UKPLAT_MEMRF_READ;
+			else
+				rt_mrd->flags |= UKPLAT_MEMRF_READ |
+						 UKPLAT_MEMRF_WRITE;
+		else
+			rt_mrd->flags |= UKPLAT_MEMRF_READ |
+					 UKPLAT_MEMRF_EXECUTE;
+
+		mat_md = (uk_efi_mem_desc_t *)((__u8 *)mat_md + desc_sz);
+	}
+}
+
+static void uk_efi_setup_bootinfo_mrds(struct ukplat_bootinfo *bi)
+{
+	struct ukplat_memregion_desc mrd = {0}, *rt_mrds;
+	uk_efi_mem_desc_t *map_start, *map_end, *md;
+	uk_efi_uintn_t map_sz, desc_sz;
+	uk_efi_status_t status;
+	__u32 rt_mrds_count, i;
+	int rc;
+
+#if defined(__X86_64__)
+	rc = ukplat_memregion_list_insert_legacy_hi_mem(&bi->mrds);
+	if (unlikely(rc < 0))
+		uk_efi_crash("Failed to insert legacy high memory region\n");
+#endif
+
+	/* Fetch the Runtime Services memory regions from the MAT */
+	uk_efi_rt_md_to_bi_mrds(&rt_mrds, &rt_mrds_count);
+	for (i = 0; i < rt_mrds_count; i++) {
+		rc = ukplat_memregion_list_insert(&bi->mrds, &rt_mrds[i]);
+		if (unlikely(rc < 0))
+			uk_efi_crash("Failed to insert rt_mrd\n");
+	}
+
+	/* We no longer need the list of Runtime Services memory regions */
+	status = uk_efi_bs->free_pool(rt_mrds);
```

Comment:
It does `memmove` though :confused:. Those are just memory regions built specifically from  the MAT and since they get `memmove`d to a different address we do not need them anymore, so I free that memory. and that memory does end up being unmapped.

Reply:
Bruh... 🤦🏼 I never asked